PT-2037: option --system-only for pt-stalk

svetasmirnova opened the pull request:
Fix for PT-2037: Add option --skip-mysql or --system-only for pt-stalk

svetasmirnova committed:
PT-2037 Added option --system-only to pt-stalk

svetasmirnova committed:
PT-2037 Option --system-only for pt-stalk

Option --system-only triggers only operating system related captures, ignoring all others.

denisok commented on bin/pt-stalk:
```diff
@@ -796,84 +797,88 @@ collect() {
-   if [ ! "$OPT_MYSQL_ONLY" ]; then
-      mysqld_pid=$(_pidof mysqld | awk '{print $1; exit;}')
-   fi
+   if [ ! "$OPT_SYSTEM_ONLY" ]; then
```
instead of `if` maybe it is good idea to create new function for system collect? and just call it in the beginning and exit if "$OPT_SYSTEM_ONLY" is set. 

And do we need new key `OPT_SYSTEM_ONLY` at all? if there is no mysqld process - just don't execute mysqld section, if OPT_MYSQL_ONLY is set don't execute the system function

svetasmirnova replied:
I made pt-stalk data collection more modular in commit d6e89940..623fdaec. Please review.

Regarding the option: while the main purpose for this patch is to allow pt-stalk to run on machines where MySQL is not installed, we may need to run it on machines where MySQL is installed, but not the source of the problem. For example, working with a customer who has a small MySQL server and large MongoDB or PostgreSQL that we are troubleshooting.

svetasmirnova committed:
PT-2037 - reviewer suggestions

Refactoring pt-stalk to make it easier to call system and MySQL data collections separately

svetasmirnova committed:
PT-2037 - option --system-only for pt-stalk

This commit:
1. Moves all changes to the proper place: lib/bash/collect.sh
2. Refactors pt-stalk so it is more modular
3. Places fix for PT-1734 into the right place: lib/bash/collect.sh
4. Fixes tests for PT-1336

svetasmirnova committed:
PT-2037 - regression fixes

Re-added fix for PT-1340 that was reverted by fix for PT-1398

denisok commented on bin/pt-stalk:
```diff
@@ -128,7 +128,6 @@ OPT_ERRS=0        # How many command line option errors
 OPT_VERSION=""    # If --version was specified
 OPT_HELP=""       # If --help was specified
 OPT_ASK_PASS=""   # If --ask-pass was specified
-OPT_MYSQL_ONLY="" # If --mysql-only was specified
```
why did you remove this option ? you added another one so you should add it here. Probably it was missed from your previous commits.

svetasmirnova replied:
Because this is part of the auto-copied code from lib/bash/parse_options.sh The tool works correctly without it

svetasmirnova replied:
Function _eval_po from parse_options.sh evaluates such options automatically and assigns value of mysql-only to OPT_MYSQL_ONLY, value system-only to OPT_SYSTEM_ONLY, and so on

denisok commented on bin/pt-stalk:
```diff
@@ -459,7 +458,6 @@ _parse_command_line() {
          else
             spec=$(grep "^short form:-$opt\$" "$PT_TMPDIR"/po/* | cut -d ':' -f 1)
             if [ -z "$spec"  ]; then
-               option_error "Unknown option: $real_opt"
```
why is this removed?

svetasmirnova replied:
This is fix for https://jira.percona.com/browse/PT-1720 that was not added to pt-stalk, because modules were not updated properly

denisok commented on bin/pt-stalk:
```diff
@@ -796,9 +801,65 @@ collect() {
+collect_mysql_data_one() {
    if [ ! "$OPT_MYSQL_ONLY" ]; then
-      mysqld_pid=$(_pidof mysqld | awk '{print $1; exit;}')
+      port=$($CMD_MYSQL $EXT_ARGV -ss -e 'SELECT @@port')
```
ok, is this some kind of improvement? Should we separate this to another ticket, so it could be more descriptive. Because adding it in the scope of this fix - it could be hard to figure out later why it was done.

svetasmirnova replied:
This is the fix for https://jira.percona.com/browse/PT-1398

denisok commented on bin/pt-stalk:
```diff
@@ -820,27 +881,27 @@ collect() {
       tail -f "$mysql_error_log" >"$d/$p-log_error" &
       tail_error_log_pid=$!
 
-      $CMD_MYSQLADMIN $EXT_ARGV debug
```
why we delete this?

svetasmirnova replied:
This is the fix for https://jira.percona.com/browse/PT-1340 that was removed by the fix for PT-1398

denisok commented on bin/pt-stalk:
```diff
@@ -1171,11 +1219,11 @@ slave_status() {
       $CMD_MYSQL $EXT_ARGV -e "$sql" >> $outfile
 
-      sql="SELECT * FROM replication_applier_status JOIN replication_applier_status_by_coordinator USING(channel_name)\G"
+      sql="SELECT * FROM performance_schema.replication_applier_status JOIN performance_schema.replication_applier_status_by_coordinator USING(channel_name)\G"
```
how did it work before ? or it will drop support for some old DB versions?

svetasmirnova replied:
It did not work. The log contained error messages that table replication_applier_status is not found. I believe this was, again, the result of mix-up of library and tool code.

denisok commented on t/pt-stalk/pt-stalk.t:
```diff
@@ -542,20 +542,68 @@ $retval = system("$trunk/bin/pt-stalk --no-stalk --run-time 2 --sleep 2 --dest $
 );
 
+# ###########################################################################
+# Test if option --system-only works correctly
```
what if both options are given ? could you add test for that and check how your system behaves ?

svetasmirnova replied:
Test case added. See commit 3767e57e151fe7e3f9631d5c736334c1b8fc431b

denisok reviewed (commented):
Overall looks good, I have couple of questions.

svetasmirnova committed:
PT-2037 - added test for the situation when both --system-only and --mysql-only are specified
